### DISCUSSION TITLE
What additional features are needed?

### DISCUSSION THREAD

**Discussion by jcassel** [jcassel/OctoPrint-Siocontrol#8] [Q&A]
Looking to understand what other features people would like to see in future releases. This was born out of my own personal use case and right now it really fits it well. That said, I know there are a lot of other things that people do with the Pi related to IO. I want to make this a real alternative for those that have chosen to use hardware that does not have native IO.

**Comment by TOBi-13** [comment]
I have been looking for something like this for a long time to have a seperate IO away from my PI.  I was just wondering if you will be adding support for using analog inputs/outputs. I think this would be a nice feature to have for example if you want to control an enclosure with temp sensors or dimmable lights. Otherwise keep up the good work.

**Comment by jcassel** [comment]
@TOBi-13 ,  I do have plans on expanding the system to have closed loop control. This would include temperature control. Which to do right needs good analog In and Out.  I am looking at the most used capabilities currently in the deprecated enclosure plugin as a starting point. 

One of the reasons this project exists is the idea that running a Pi or multiple Pi located at the printers is not very reliable long term or cost effective.  I have a lot to say on this subject but I will keep it to myself for now. :) 

You might also like that I am very close to putting out a version that is not just serial connected but can be optionally controlled over a network connection... Wifi or hard line(dependent on the controller / board) . I prefer a hard line for stability of connection but Wifi is a no brainer. Many users will think that network control right now does not really add too to the echo system given the wifi power switches that there are other plugins for. The power of this is the inputs and the fact that you could have a completely off loaded process like PID control or sequence control that works outside the idea of OctoPrint yet you can use the OctoPrint interface to stitch it all together and monitor it. 

As I respond here I am realizing just how many plans I have. I am going to put up a road map in the Wiki to see if users have an opinion on what is most important in my plans. 

Thanks for the feedback.  Stay tuned there is a lot to come.

**Comment by mikaelsundin** [comment]
G-code parsing would be nice, for example control external fans.

Good work jcassel btw 😃

**Reply by mikaelsundin** [reply]
Found this on the topic.
https://github.com/b-morgan/OctoPrint-FanSpeedMirror/blob/master/octoprint_FanSpeedMirror/__init__.py

**Reply by jcassel** [reply]
@mikaelsundin Thanks for the suggestion.  Seems like something that could be made part of the action control scripts update coming. This would make sense to add control for any IO point from GCode.

**Comment by DIY89** [comment]
Like Tobi i'm searching a lot for a plugin like this to seperate the IO-Control to a external IO-Board (eg. ardunino). I already soldered an Arduino HAT Board to read sensor data (DHT22), check IOs (door sensor, smoke detector), control fan, relais etc. Do you have an software for example arduino side? I'm curious to see more.

**Reply by jcassel** [reply]
@DIY89 I am not sure I understand the question fully but... 
The firmware for this project is available and it does have an Arduino version. 
As for reading a DHT22, I am a few weeks away from releasing an update to the ESP32 firmware and a sub-Plugin that reads up to 4 DHT22 devices and can tie the reading to an output for environmental controls. This sub plugin will be labeled as SIOEnclosure. Temp control bill be bang bang in the first release but I will add support for PWM as a follow up.

I am done with the Device side of PWM output for things like a heater or fan control. The serial AIP completed but I have not started on exposing that in OctoPrint. This will come after what I mentioned above with the DHT22 sensor.  

Oh and I should mention that I have a sub plugin coming out that allow you to configure the IO types from within OctoPrint.  This will come out around the same time as the DHT Sensor reading and temp control Sub-Plugin. 

Hope that helps. Let me know if there is something I am missing.

**Reply by DIY89** [reply]
Now i get it. I'd just take a look in the `https://github.com/jcassel/OctoPrint_SIOControl_Firmware/blob/main/SIO_Arduino_Minimal_Base/SIO_Arduino_Minimal_Base.ino`. That is were i was looking for. Great job.
For future features i was thinking about to add i²c stuff like BME280 sensors, port extendors ....

**Reply by jcassel** [reply]
Ya, I have also been thinking I2C is something that needs to be implemented. It's on the list..
